fabric-network/src/impl/query/queryagent.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2020 IBM All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict';
+
+const logger = require('fabric-network/lib/logger').getLogger('QueryAgent');
+
+/**
+ * @typedef {Object} QueryAgent~QueryResponse
+ * @memberof module:fabric-network
+ * @property {number} status - The status value from the endorsement. This attriibute
+ * will be set by the chaincode.
+ * @property {Buffer} payload - The payload value from the endorsement. This attribute
+ * may be considered the query value if the status code is exceptable.
+ * @property {Buffer} payload - The message value from the endorsement. This attribute
+ * will have a value when there is not a payload and status value indicates an issue
+ * with determining the payload (the query value).
+ */
+/**
+ * Used by query handler implementations to evaluate transactions on peers of their choosing.
+ * @memberof module:fabric-network
+ */
+class QueryAgent {
+	/**
+	 * Builds a QueryAgent instance to send and then work with the results returned
+	 * by the fabric-common/Query.
+	 * @param {module:fabric-common.Query} query - The query instance of the proposal
+	 * @returns {Object} options - options to be used when sending the request to
+	 * fabric-common service endpoint {Endorser} peer.
+	 */
+	constructor(query, options = {}) {
+		this.query = query;
+		this.requestTimeout = 3000; // default 3 seconds
+		if (Number.isInteger(options.timeout)) {
+			this.requestTimeout = options.timeout * 1000; // need ms;
+		}
+	}
+
+	/**
+	 * Sends a signed proposal to the specified peers. The peer endorsment
+	 * responses are
+	 * @param {Endorser[]} peers - The peers to query
+	 * @returns {Object.<String, (QueryResponse | Error)>} Object with peer name keys and associated values that are either
+	 * QueryResponse objects or Error objects.
+	 */
+	async send(peers) {
+		const method = 'send';
+		logger.debug('%s - start', method);
+
+		const results = {};
+		try {
+			const responses = await this.query.send({targets: peers, requestTimeout: this.requestTimeout});
+			if (responses) {
```

Comment:
If the send can return no results, does this mean any arbitrary subset of peer results (or errors) might be missing? If only a subset are missing then I don't think this code ensures that the return value is always an object with a property for each target peer name.

Reply:
Added a check to be sure all peers requested have a response